File: Server/src/game/common/security/AESUtil.java
package game.common.security;

import org.bouncycastle.pqc.math.linearalgebra.ByteUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.crypto.Cipher;
import javax.crypto.KeyGenerator;
import javax.crypto.SecretKey;
import javax.crypto.spec.IvParameterSpec;
import java.security.NoSuchAlgorithmException;
import java.security.SecureRandom;

/**
 * AES加密和解密工具
 * Created by wuy on 2017/5/22.
 */
public class AESUtil {

	final private static Logger logger = LoggerFactory.getLogger(AESUtil.class);
	final private static String ALGORITHM = "AES";
	final private static String ALGORITHM_MODE = "AES/CBC/PKCS5Padding";
	final private static int KEY_LENGTH = 128;
	final private static IvParameterSpec IV = new IvParameterSpec("1q2w3e4r5t6y7u8i".getBytes());

	/**
	 * 产生随机密钥
	 *
	 * @return 密钥
	 */
	public static SecretKey createSecretKey() {
		try {
			KeyGenerator keygen = KeyGenerator.getInstance(ALGORITHM);
			keygen.init(KEY_LENGTH, new SecureRandom());
			return keygen.generateKey();
		} catch (NoSuchAlgorithmException e) {
			e.printStackTrace();
		}
		return null;
	}

	/**
	 * 加密
	 *
	 * @param plainBytes 明文
	 * @return 密文
	 */
	public static byte[] encrypt(byte[] plainBytes, SecretKey key) {
		try {
			Cipher cipher = Cipher.getInstance(ALGORITHM_MODE);
			cipher.init(Cipher.ENCRYPT_MODE, key, IV);
			return cipher.doFinal(plainBytes);
		} catch (Exception e) {
			logger.error("encrypt error:", e);
		}
		return null;
	}

	/**
	 * 解密
	 *
	 * @param cipherBytes 密文
	 * @return 明文
	 */
	public static byte[] decrypt(byte[] cipherBytes, SecretKey key) {
		try {
			Cipher cipher = Cipher.getInstance(ALGORITHM_MODE);
			cipher.init(Cipher.DECRYPT_MODE, key, IV);
			return cipher.doFinal(cipherBytes);
		} catch (Exception e) {
			logger.error("decrypt error:", e);
		}
		return null;
	}

	public static void main(String[] args) throws NoSuchAlgorithmException {
		SecretKey key = AESUtil.createSecretKey();
		System.out.println("SecretKey:" + ByteUtils.toHexString(key.getEncoded()));

		byte[] plainBytes = ByteUtils.fromHexString("83a169a56c6f67696ea170b882a76163636f756e74a3313233a870617373776f7264a131a163ad66726f6d5f6c735f6c6f67696e");
//		byte[] plainBytes = "HelloWorld哈哈".getBytes();
		System.out.println("plainBytes length:" + plainBytes.length + ", plainBytes:" + ByteUtils.toHexString(plainBytes));

		byte[] cipherBytes = AESUtil.encrypt(plainBytes, key);
		System.out.println("cipherBytes length:" + cipherBytes.length + ", cipherBytes:" + ByteUtils.toHexString(cipherBytes));

		plainBytes = AESUtil.decrypt(cipherBytes, key);
		System.out.println("plainBytes length:" + plainBytes.length + ", plainBytes:" + ByteUtils.toHexString(plainBytes));
	}
}

File: Client/game/client/security/aes_util.py
# coding=utf-8
from binascii import b2a_hex, a2b_hex

from Crypto.Cipher import AES


class AESUtil:
	"""
	AES加密和解密工具，速度优于RSA
	"""

	def __init__(self):
		self._key = None
		self._iv = '1q2w3e4r5t6y7u8i'

	def set_key(self, aes_key):
		"""
		设置密钥
		:param aes_key: 密钥
		"""
		self._key = a2b_hex(aes_key)

	def fill_plain_bytes(self, plain_bytes):
		"""
		填充明文，保证明文长度是AES.block_size的整数倍(Java层AES/CBC/PKCS5Padding)
		:param plain_bytes: 明文
		:return: 填充后的明文
		"""
		fill_bytes_length = AES.block_size - len(plain_bytes) % AES.block_size
		if fill_bytes_length == 0:
			fill_bytes_length = AES.block_size
		return plain_bytes + fill_bytes_length * chr(fill_bytes_length)

	def encrypt(self, plain_bytes):
		"""
		加密
		:param plain_bytes: 明文
		:return: 密文
		"""
		cipher = AES.new(self._key, AES.MODE_CBC, self._iv)
		return cipher.encrypt(self.fill_plain_bytes(plain_bytes))

	def decrypt(self, cipher_bytes):
		"""
		解密
		:param cipher_bytes: 密文
		:return: 明文
		"""
		cipher = AES.new(self._key, AES.MODE_CBC, self._iv)
		plain_bytes = cipher.decrypt(cipher_bytes)
		return plain_bytes.rstrip(plain_bytes[-1])


if __name__ == '__main__':
	aes_util = AESUtil()
	aes_util.set_key('b50af4b6261389dfe1b6ca87a868c324')

	# plain_bytes = 'HelloWorld哈哈'
	plain_bytes = a2b_hex('83a169a56c6f67696ea170b882a76163636f756e74a3313233a870617373776f7264a131a163ad66726f6d5f6c735f6c6f67696e')
	print 'plain_bytes length:', len(plain_bytes), 'plain_bytes:', b2a_hex(plain_bytes)

	cipher_bytes = aes_util.encrypt(plain_bytes)
	print 'cipher_bytes length:', len(cipher_bytes), 'cipher_bytes:', b2a_hex(cipher_bytes)

	plain_bytes = aes_util.decrypt(cipher_bytes)
	print 'plain_bytes length:', len(plain_bytes), 'plain_bytes:', b2a_hex(plain_bytes)
